felix/fv/bpf_notracking_test.go

```diff
@@ -0,0 +1,156 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fv_test
+
+import (
+	"fmt"
+	"os"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"github.com/pkg/errors"
+	"github.com/vishvananda/netlink"
+
+	"github.com/projectcalico/calico/felix/fv/connectivity"
+	"github.com/projectcalico/calico/felix/fv/containers"
+	"github.com/projectcalico/calico/felix/fv/infrastructure"
+	"github.com/projectcalico/calico/felix/fv/workload"
+	"github.com/projectcalico/calico/libcalico-go/lib/apiconfig"
+)
+
+var _ = infrastructure.DatastoreDescribe("_BPF-SAFE_ Felix bpf with NOTRACK feature", []apiconfig.DatastoreType{apiconfig.EtcdV3}, func(getInfra infrastructure.InfraFactory) {
+
+	if os.Getenv("FELIX_FV_ENABLE_BPF") != "true" {
+		// Non-BPF run.
+		return
+	}
+
+	var (
+		infra          infrastructure.DatastoreInfra
+		felixes        []*infrastructure.Felix
+		workloads      [2]*workload.Workload
+		cc             *connectivity.Checker
+		externalClient *containers.Container
+	)
+
+	BeforeEach(func() {
+		infra = getInfra()
+
+		opts := infrastructure.DefaultTopologyOptions()
+		opts.ExtraEnvVars = map[string]string{
+			"FELIX_BPFENABLED":              "true",
+			"FELIX_BPFHostConntrackBypass":  "true",
+			"FELIX_DEBUGDISABLELOGDROPPING": "true"}
+
+		felixes, _ = infrastructure.StartNNodeTopology(2, opts, infra)
+		infra.AddDefaultAllow()
+		cc = &connectivity.Checker{}
+
+		// Wait until the tunl0 device appears; it is created when felix inserts the ipip module
+		// into the kernel.
+		Eventually(func() error {
+			links, err := netlink.LinkList()
+			if err != nil {
+				return err
+			}
+			for _, link := range links {
+				if link.Attrs().Name == "tunl0" {
+					return nil
+				}
+			}
+			return errors.New("tunl0 wasn't auto-created")
+		}).Should(BeNil())
+
+		for i := range felixes {
+			const portsToOpen = "80"
+			workloads[i] = workload.Run(felixes[i],
+				fmt.Sprintf("host%d-webserver", i),
+				"default",
+				fmt.Sprintf("10.65.%d.2", i),
+				portsToOpen, "tcp")
+			workloads[i].ConfigureInInfra(infra)
+		}
+
+		externalClient = infrastructure.RunExtClient("ext-client")
+	})
+
+	AfterEach(func() {
+		if CurrentGinkgoTestDescription().Failed {
+			infra.DumpErrorData()
+		}
+
+		for _, felix := range felixes {
+			felix.Stop()
+		}
+
+		infra.Stop()
+	})
+
+	It("should allow 3rd party DNAT to workloads work", func() {
+		expectNormalConnectivity := func() {
+			cc.ResetExpectations()
+			for i := range felixes {
+				cc.ExpectSome(felixes[i], workloads[0].Port(80))
```

Comment:
I think the problem was more like workload->host then redirected to workload by host.

So I think there should be first Expect(None, workloads[0], felixs[i]. port) then Expect(Some...) after the rules are added then Expect(None,...) again.

I wonder if it would be simpler to add this in the bpf_test.go as we can then run it in various overlay scenarios.

Reply:
This line is just to test if normal connectivity is OK, and the change is not breaking something that should always work.
The test that is expected to change is 
```hostIP := connectivity.TargetIP(felixes[i].IP)
   cc.ExpectNone(externalClient, hostIP, 8080)```

Reply:
> This line is just to test if normal connectivity is OK, and the change is not breaking something that should always work.

Don't the other tests take care of that. Perhaps it would make sense to add this _after_ the DNAT rule is injected?

I think we are really interested here only in pod->host or extclient->host cases. Those are the cases that do not work. And afaict, they are not supposed to work before the the DNAT rule is injected because there should be no host networked backend (ideally).

Reply:
Ok, I removed those tests.